### PR TITLE
TreeNamingService

### PR DESCRIPTION
**Pull request by BeckmaR**
Added a new NamingService that adds all given names to a tree before it shortens the names.
Also new are the classes ShortString, StringTreeNode and ElementNameProvider.

This was created as a possible solution for issue #194.

### CODE REVIEW

File: plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/naming/ElementNameProvider.xtend [hunk 1]
```diff
@@ -0,0 +1,78 @@
+package org.yakindu.sct.model.sexec.naming
```
Comment (andreasmuelder) [inline plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/naming/ElementNameProvider.xtend]: Please add EPL licence header and author tag to all classes

File: plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/naming/ShortString.xtend [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+	// cost of cutting operations
+	final static public int COST_LOWERCASE_VOCALS = 1;
+	final static public int COST_UNDERSCORE = 1;
+	final static public int COST_LOWERCASE_CONSONANTS = 2;
+	final static public int COST_UPPERCASE = 3;
+	final static public int COST_DIGIT = 10;
+	final static public int COST_FIRSTLETTER = 10;
+	
+	
+	new(String s)
```
Comment (andreasmuelder) [inline plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/naming/ShortString.xtend]: Please reformat classes with defined Eclipse Formatter settings

File: plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/naming/TreeNamingService.xtend [hunk 3]
```diff
@@ -0,0 +1,695 @@
...
+import com.google.common.collect.Maps
+
+/** Default implementation of the naming service for various identifiers used in the generated code. 
+ * It is responsible for identifier construction depending on the thing to be named including different strategies 
+ * which also include name shortening.
+ */
+class TreeNamingService implements INamingService {
+
+	@Inject extension SExecExtensions
+	//@Inject extension SgraphExtensions
```
Comment (andreasmuelder) [inline plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/naming/TreeNamingService.xtend]: Remove code that is commented out

File: plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/naming/TreeNamingService.xtend [hunk 4]
```diff
@@ -0,0 +1,695 @@
...
+		{
+			if (executionState.sourceElement instanceof State) 
+			{
+				val state = executionState.sourceElement as State
+				addShortTimeEventName(executionState, state)
+			}
+		}
+	}
+	
+	def public test_printTreeContents()
```
Comment (andreasmuelder) [inline plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/naming/TreeNamingService.xtend]: Is this still needed?

File: test-plugins/org.yakindu.sct.model.sexec.test/src/org/yakindu/sct/model/sexec/transformation/test/TreeNamingServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+						System.out.println(n);
+					}
+				}
+				assertEquals(names.contains(name), false);
+				names.add(name);
+				System.out.println(name);
+			}
+			System.out.println();
+//			SExecExtensions ext = new SExecExtensions();
+//			for(Step step : ext.getAllFunctions(flow))
```
Comment (andreasmuelder) [inline test-plugins/org.yakindu.sct.model.sexec.test/src/org/yakindu/sct/model/sexec/transformation/test/TreeNamingServiceTest.java]: remove commented code

File: test-plugins/org.yakindu.sct.test.models/src/org/yakindu/sct/test/models/SCTUnitTestModels.java [hunk 6]
```diff
@@ -1,125 +1,126 @@
-/**
```
Comment (andreasmuelder) [inline test-plugins/org.yakindu.sct.test.models/src/org/yakindu/sct/test/models/SCTUnitTestModels.java]: Something went wrong with the diff of SCTUnitTestModels maybe due to reformatting?

Comment (BeckmaR) [inline test-plugins/org.yakindu.sct.test.models/src/org/yakindu/sct/test/models/SCTUnitTestModels.java]: Don't know what happened, I only inserted one line or something like that.

File: plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/naming/TreeNamingService.xtend (hunk 4 above)
Comment (BeckmaR) [inline plugins/org.yakindu.sct.model.sexec/src/org/yakindu/sct/model/sexec/naming/TreeNamingService.xtend]: Yes, why not?

**Comment by andreasmuelder** [comment]
@BeckmaR @terfloth Please decide if this will be merged to master

**Comment by jdicks** [comment]
@BeckmaR @terfloth Please decide if this will be merged to master :)

**Pull request merged**